src/test/scala/rise/autotuning.scala

```diff
@@ -119,6 +119,37 @@ class autotuning extends test_util.Tests {
   }
   // scalastyle:on
 
+
+  // function to check cycle in parameter dependencies
+  def check_no_cycle(distribution: Map[NatIdentifier,
+    (Set[constraints.Constraint], Set[NatIdentifier])]
+                    ): Boolean = {
+
+    def check_no_cycle_rec(param: NatIdentifier,
+                           dependencies: Set[NatIdentifier],
+                           distribution: Map[NatIdentifier,
+                             (Set[constraints.Constraint], Set[NatIdentifier])]
+                          ): Boolean = {
+      dependencies.size match {
+        case 0 => true
+        case _ => {
+          dependencies.exists(dependency => dependency.name.equals(param.name)) match {
+            case true => false
+            case false => {
+              dependencies.exists(dependency => {
+                check_no_cycle_rec(param, distribution.apply(dependency)._2, distribution)
+              })
+            }
+          }
+        }
+      }
+    }
+
+    distribution.exists(param => {
+      check_no_cycle_rec(param._1, param._2._2, distribution)
+    })
```

Comment:
I am not sure to understand your algorithm here. There is no cycle if there exists a parameter which does not induce a cycle? Would this example pass the check:
```
A depends on B
B depends on A
C depends on D
```
I am also surprised that you do not use a set of visited nodes/parameters in your algorithm.

Reply:
You are right. This should not work properly. I updated the algorithm.

Reply:
The algorithm is updated.